clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -552,9 +552,11 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
         if (response.wasDisconnected()) {
             log.trace("Cancelled request with header {} due to node {} being disconnected",
                 requestHeader, response.destination());
-            for (ProducerBatch batch : batches.values())
-                completeBatch(batch, new ProduceResponse.PartitionResponse(Errors.NETWORK_EXCEPTION, String.format("Disconnected from node %s", response.destination())),
+            for (ProducerBatch batch : batches.values()) {
+                Errors e = response.wasTimedOut() ? Errors.REQUEST_TIMED_OUT : Errors.NETWORK_EXCEPTION;
```

Comment:
I wonder if propagating the TimeoutException to the user space has any backward compatibility implications. I think that we already use TimeoutException when the batch expires so the user should already handle timeouts. That seems to be ok.